File: javascript/sandbox-setup.js
#!/usr/bin/env node
// sandbox-setup.sh — portiert nach javascript
// Quelle: shell, Onboarding@main:scripts/sandbox-setup.sh
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

// Provisioniert die Claude-Code-Sandbox (Remote-Umgebung) reproduzierbar:
//   - Node-Dependencies (Frontend, npm)
//   - Python-Dependencies (Backend inkl. pytest)
//   - Medien-Tools: ffmpeg, ImageMagick, GIMP, Blender headless (apt) —
//     Fehlschlag blockiert die Session nicht; --skip-heavy laesst GIMP/Blender aus
// Idempotent: bereits Vorhandenes wird uebersprungen; der Container-Cache der
// Umgebung macht die apt-Installation zum Einmal-Aufwand.

import { execSync, spawn } from 'child_process';
import { writeFileSync, existsSync, readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const args = process.argv.slice(2);
const SKIP_HEAVY = args.includes('--skip-heavy');

process.chdir(join(__dirname, '..'));

function log(message) {
  console.log(`[sandbox-setup] ${message}`);
}

function runCommand(command, options = {}) {
  try {
    const result = execSync(command, { encoding: 'utf8', ...options });
    return result.trim();
  } catch (error) {
    return null;
  }
}

function commandExists(command) {
  return runCommand(`which ${command}`) !== null;
}

log('Node-Dependencies (npm install) …');
try {
  execSync('npm install --no-audit --no-fund', { stdio: 'inherit' });
} catch (error) {
  log('FEHLER: npm install fehlgeschlagen');
  process.exit(1);
}

log('Python-Dependencies (backend/requirements-dev.txt) …');
try {
  execSync('pip3 install --quiet -r backend/requirements-dev.txt', { stdio: 'inherit' });
} catch (error) {
  log('FEHLER: pip install fehlgeschlagen');
  process.exit(1);
}

let APT_UPDATED = false;

function aptInstall(pkg, bin) {
  if (commandExists(bin)) {
    const versionOutput = runCommand(`${bin} -version 2>&1 | head -1`) || '';
    log(`${pkg} bereits vorhanden (${versionOutput})`);
    return true;
  }
  
  log(`Installiere ${pkg} …`);
  try {
    if (!APT_UPDATED) {
      execSync('DEBIAN_FRONTEND=noninteractive apt-get update -qq', { stdio: 'pipe' });
      APT_UPDATED = true;
    }
    execSync(`DEBIAN_FRONTEND=noninteractive apt-get install -y -qq ${pkg}`, { stdio: 'pipe' });
    return true;
  } catch (error) {
    log(`WARNUNG: ${pkg} konnte nicht installiert werden (Netzwerk-Policy?) — Medien-Schritte ggf. eingeschraenkt`);
    return false;
  }
}

aptInstall('ffmpeg', 'ffmpeg');
aptInstall('imagemagick', 'convert');

if (!SKIP_HEAVY) {
  aptInstall('gimp', 'gimp');
  aptInstall('blender', 'blender');
}

aptInstall('xvfb', 'Xvfb');
aptInstall('x11-utils', 'xdpyinfo');
aptInstall('libnss3-tools', 'certutil');

if (!commandExists('google-chrome-stable')) {
  log('Installiere Google Chrome Stable …');
  const tmpDeb = '/tmp/google-chrome.deb';
  try {
    execSync(`curl -fsSL -o ${tmpDeb} https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb`, { stdio: 'pipe' });
    execSync(`DEBIAN_FRONTEND=noninteractive apt-get install -y -qq ${tmpDeb}`, { stdio: 'pipe' });
    const chromeVersion = runCommand('google-chrome-stable --version') || '';
    log(`Chrome installiert: ${chromeVersion}`);
  } catch (error) {
    log('WARNUNG: Chrome-Installation fehlgeschlagen');
  }
}

// Proxy-CA in Chromes NSS-DB, damit externes HTTPS ohne Zertifikatsfehler läuft.
if (commandExists('certutil') && existsSync('/root/.ccr/ca-bundle.crt')) {
  const nssdbPath = `${process.env.HOME}/.pki/nssdb`;
  try {
    execSync(`mkdir -p ${nssdbPath}`, { stdio: 'pipe' });
    execSync(`certutil -d sql:${nssdbPath} -N --empty-password`, { stdio: 'pipe' });
    
    const certList = runCommand(`certutil -d sql:${nssdbPath} -L`) || '';
    if (!certList.includes('ccr-proxy-ca')) {
      execSync(`certutil -d sql:${nssdbPath} -A -t "C,," -n ccr-proxy-ca -i /root/.ccr/ca-bundle.crt`, { stdio: 'pipe' });
      log('Proxy-CA in Chrome-NSS-Store importiert');
    }
  } catch (error) {
    // Ignore errors
  }
}

// Playwright-Node-Module ins Projekt verlinken (visual-qa.mjs / browser-session.mjs).
if (existsSync('node_modules') && !existsSync('node_modules/playwright')) {
  if (commandExists('npm')) {
    try {
      execSync('npm install --no-audit --no-fund --no-save playwright', { stdio: 'pipe' });
      log('Playwright (Node) installiert');
    } catch (error) {
      log('WARNUNG: Playwright-npm-Install fehlgeschlagen');
    }
  }
}

// Git-Push-Weg: Der Session-Git-Proxy (origin) ist read-only. Pushes laufen
// direkt zu github.com mit dem Nutzer-PAT (GH_ACCESS_TOKEN aus Umgebungs-Env
// oder .env, geliefert vom Credential-Helper — kein Secret in der Git-Config).
try {
  execSync('git rev-parse --is-inside-work-tree', { stdio: 'pipe' });
  const pwd = process.cwd();
  execSync(`git config credential."https://x-access-token@github.com".helper "!${pwd}/.claude/git-credential-pat.sh"`);
  execSync('git remote set-url --push origin "https://x-access-token@github.com/KikiKari/Onboarding.git"');
  log('Git-Push-Route: direkt zu github.com (PAT via Credential-Helper)');
} catch (error) {
  // Not a git repository or other git error
}

// Docker-Daemon fuer Dev-Compose-Verifikation in der Sandbox.
// Docker-Hub-Blobs (cloudfront.docker.com) sind von der Netz-Policy blockiert —
// mirror.gcr.io liefert die Library-Images. Container brauchen zusaetzlich die
// Proxy-CA (siehe docker-compose.sandbox.yml).
if (commandExists('dockerd') && !runCommand('docker info')) {
  log('Starte Docker-Daemon (Registry-Mirror: mirror.gcr.io) …');
  
  try {
    execSync('mkdir -p /etc/docker', { stdio: 'pipe' });
    if (!existsSync('/etc/docker/daemon.json')) {
      writeFileSync('/etc/docker/daemon.json', JSON.stringify({
        'registry-mirrors': ['https://mirror.gcr.io']
      }));
    }
    
    // Start docker daemon in background
    const dockerProcess = spawn('dockerd', [], {
      stdio: ['ignore', '/tmp/dockerd.log', '/tmp/dockerd.log'],
      detached: true
    });
    dockerProcess.unref();
    
    // Wait for docker to start
    let dockerStarted = false;
    for (let i = 0; i < 15; i++) {
      if (runCommand('docker info')) {
        dockerStarted = true;
        break;
      }
      execSync('sleep 1');
    }
    
    if (dockerStarted) {
      log('Docker-Daemon laeuft');
    } else {
      log('WARNUNG: Docker-Daemon nicht gestartet');
    }
  } catch (error) {
    log('WARNUNG: Docker-Daemon nicht gestartet');
  }
}

log('Fertig. Versionen:');
console.log('[sandbox-setup]   node ' + (runCommand('node --version') || ''));
console.log('[sandbox-setup]   ' + (runCommand('python3 --version') || ''));

if (commandExists('ffmpeg')) {
  console.log('[sandbox-setup]   ' + (runCommand('ffmpeg -version 2>/dev/null | head -1') || ''));
}
if (commandExists('convert')) {
  console.log('[sandbox-setup]   ' + (runCommand('convert -version 2>/dev/null | head -1') || ''));
}
if (commandExists('gimp')) {
  console.log('[sandbox-setup]   ' + (runCommand('gimp --version 2>/dev/null | head -1') || ''));
}
if (commandExists('blender')) {
  console.log('[sandbox-setup]   ' + (runCommand('blender --version 2>/dev/null | head -1') || ''));
}

process.exit(0);

File: python/sandbox-setup.py
#!/usr/bin/env python3
# sandbox-setup.sh — portiert nach python
# Quelle: shell, Onboarding@main:scripts/sandbox-setup.sh
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

import os
import sys
import subprocess
import shutil
import tempfile
import time

def main():
    skip_heavy = len(sys.argv) > 1 and sys.argv[1] == "--skip-heavy"
    
    os.chdir(os.path.join(os.path.dirname(__file__), ".."))
    
    def log(message):
        print(f"[sandbox-setup] {message}")
    
    log("Node-Dependencies (npm install) …")
    try:
        subprocess.run(["npm", "install", "--no-audit", "--no-fund"], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except subprocess.CalledProcessError:
        log("FEHLER: npm install fehlgeschlagen")
        sys.exit(1)
    
    log("Python-Dependencies (backend/requirements-dev.txt) …")
    try:
        subprocess.run(["pip3", "install", "--quiet", "-r", "backend/requirements-dev.txt"], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except subprocess.CalledProcessError:
        log("FEHLER: pip install fehlgeschlagen")
        sys.exit(1)
    
    apt_updated = False
    
    def apt_install(pkg, binary):
        nonlocal apt_updated
        if shutil.which(binary):
            try:
                version_output = subprocess.check_output([binary, "-version"], stderr=subprocess.STDOUT, text=True)
                version_line = version_output.split('\n')[0] if version_output else ""
            except subprocess.CalledProcessError:
                version_line = ""
            log(f"{pkg} bereits vorhanden ({version_line})")
            return
        
        log(f"Installiere {pkg} …")
        if not apt_updated:
            env = os.environ.copy()
            env["DEBIAN_FRONTEND"] = "noninteractive"
            try:
                subprocess.run(["apt-get", "update", "-qq"], check=True, env=env, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
                apt_updated = True
            except subprocess.CalledProcessError:
                pass
        
        env = os.environ.copy()
        env["DEBIAN_FRONTEND"] = "noninteractive"
        try:
            subprocess.run(["apt-get", "install", "-y", "-qq", pkg], check=True, env=env, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        except subprocess.CalledProcessError:
            log(f"WARNUNG: {pkg} konnte nicht installiert werden (Netzwerk-Policy?) — Medien-Schritte ggf. eingeschraenkt")
    
    apt_install("ffmpeg", "ffmpeg")
    apt_install("imagemagick", "convert")
    if not skip_heavy:
        apt_install("gimp", "gimp")
        apt_install("blender", "blender")
    
    apt_install("xvfb", "Xvfb")
    apt_install("x11-utils", "xdpyinfo")
    apt_install("libnss3-tools", "certutil")
    
    if not shutil.which("google-chrome-stable"):
        log("Installiere Google Chrome Stable …")
        with tempfile.NamedTemporaryFile(suffix=".deb", delete=False) as tmp_deb:
            tmp_deb_path = tmp_deb.name
        try:
            subprocess.run(["curl", "-fsSL", "-o", tmp_deb_path, "https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb"], check=True)
            env = os.environ.copy()
            env["DEBIAN_FRONTEND"] = "noninteractive"
            result = subprocess.run(["apt-get", "install", "-y", "-qq", tmp_deb_path], env=env, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            if result.returncode == 0:
                try:
                    chrome_version = subprocess.check_output(["google-chrome-stable", "--version"], stderr=subprocess.DEVNULL, text=True).strip()
                    log(f"Chrome installiert: {chrome_version}")
                except subprocess.CalledProcessError:
                    pass
            else:
                log("WARNUNG: Chrome-Installation fehlgeschlagen")
        except subprocess.CalledProcessError:
            log("WARNUNG: Chrome-Download fehlgeschlagen (Netzwerk-Policy?)")
        finally:
            if os.path.exists(tmp_deb_path):
                os.remove(tmp_deb_path)
    
    if shutil.which("certutil") and os.path.isfile("/root/.ccr/ca-bundle.crt"):
        nssdb_path = os.path.expanduser("~/.pki/nssdb")
        os.makedirs(nssdb_path, exist_ok=True)
        try:
            subprocess.run(["certutil", "-d", f"sql:{nssdb_path}", "-N", "--empty-password"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        except subprocess.CalledProcessError:
            pass
        
        try:
            output = subprocess.check_output(["certutil", "-d", f"sql:{nssdb_path}", "-L"], stderr=subprocess.DEVNULL, text=True)
            if "ccr-proxy-ca" not in output:
                result = subprocess.run(["certutil", "-d", f"sql:{nssdb_path}", "-A", "-t", "C,,", "-n", "ccr-proxy-ca", "-i", "/root/.ccr/ca-bundle.crt"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
                if result.returncode == 0:
                    log("Proxy-CA in Chrome-NSS-Store importiert")
        except subprocess.CalledProcessError:
            pass
    
    if os.path.isdir("node_modules") and not os.path.islink("node_modules/playwright") and not os.path.exists("node_modules/playwright"):
        if shutil.which("npm"):
            try:
                subprocess.run(["npm", "install", "--no-audit", "--no-fund", "--no-save", "playwright"], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
                log("Playwright (Node) installiert")
            except subprocess.CalledProcessError:
                log("WARNUNG: Playwright-npm-Install fehlgeschlagen")
    
    try:
        if subprocess.run(["git", "rev-parse", "--is-inside-work-tree"], capture_output=True, check=True).returncode == 0:
            script_dir = os.path.abspath(os.path.dirname(__file__))
            project_root = os.path.dirname(script_dir)
            credential_helper_script = os.path.join(project_root, ".claude", "git-credential-pat.sh")
            subprocess.run(["git", "config", f"credential.https://x-access-token@github.com.helper", f"!{credential_helper_script}"])
            subprocess.run(["git", "remote", "set-url", "--push", "origin", "https://x-access-token@github.com/KikiKari/Onboarding.git"])
            log("Git-Push-Route: direkt zu github.com (PAT via Credential-Helper)")
    except (subprocess.CalledProcessError, FileNotFoundError):
        pass
    
    if shutil.which("dockerd") and not is_docker_running():
        log("Starte Docker-Daemon (Registry-Mirror: mirror.gcr.io) …")
        daemon_config_path = "/etc/docker/daemon.json"
        os.makedirs(os.path.dirname(daemon_config_path), exist_ok=True)
        if not os.path.exists(daemon_config_path):
            with open(daemon_config_path, "w") as f:
                f.write('{"registry-mirrors":["https://mirror.gcr.io"]}\n')
        
        subprocess.Popen(["dockerd"], stdout=open("/tmp/dockerd.log", "w"), stderr=subprocess.STDOUT)
        
        for _ in range(15):
            if is_docker_running():
                break
            time.sleep(1)
        
        if is_docker_running():
            log("Docker-Daemon laeuft")
        else:
            log("WARNUNG: Docker-Daemon nicht gestartet")
    
    log("Fertig. Versionen:")
    try:
        node_version = subprocess.check_output(["node", "--version"], text=True).strip()
        print(f"[sandbox-setup]   node {node_version}")
    except (subprocess.CalledProcessError, FileNotFoundError):
        pass
    
    try:
        python_version = subprocess.check_output(["python3", "--version"], text=True).strip()
        print(f"[sandbox-setup]   {python_version}")
    except (subprocess.CalledProcessError, FileNotFoundError):
        pass
    
    binaries = [
        ("ffmpeg", ["ffmpeg", "-version"]),
        ("convert", ["convert", "-version"]),
        ("gimp", ["gimp", "--version"]),
        ("blender", ["blender", "--version"])
    ]
    
    for name, cmd in binaries:
        if shutil.which(name):
            try:
                version_output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, text=True)
                version_line = version_output.split('\n')[0] if version_output else ""
                print(f"[sandbox-setup]   {version_line}")
            except subprocess.CalledProcessError:
                pass

def is_docker_running():
    try:
        subprocess.run(["docker", "info"], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False

if __name__ == "__main__":
    main()
